tests/spin/test_lattice.py

```diff
@@ -0,0 +1,541 @@
+    vectors = [[0, 1], [1, 0]]
+    n_cells = [2, 2]
+    positions = [0, 0]
+    with pytest.raises(ValueError, match="The dimensions of positions array must be 2, got 1."):
+        Lattice(n_cells=n_cells, vectors=vectors, positions=positions)
+
+
+def test_vectors_shape_error():
+    r"""Test that an error is raised if a wrong dimension is entered for vectors."""
+    vectors = [[0, 1, 2], [0, 1, 1]]
+    n_cells = [2, 2]
+    with pytest.raises(ValueError, match="The number of primitive vectors must match their length"):
+        Lattice(n_cells=n_cells, vectors=vectors)
+
+
+@pytest.mark.parametrize(("n_cells"), [([2, -2]), ([2, 2.4])])
+def test_n_cells_type_error(n_cells):
+    r"""Test that an error is raised if length of vectors is provided not as an int."""
+
+    vectors = [[0, 1], [1, 0]]
+    with pytest.raises(TypeError, match="Argument `n_cells` must be a list of positive integers"):
+        Lattice(n_cells=n_cells, vectors=vectors)
+
+
+@pytest.mark.parametrize(
+    # expected_points here were calculated manually.
+    ("vectors", "positions", "n_cells", "expected_points"),
+    [
+        (
+            [[0, 1], [1, 0]],
+            [[1.5, 1.5]],
+            [3, 3],
+            [
+                [1.5, 1.5],
+                [2.5, 1.5],
+                [3.5, 1.5],
+                [1.5, 2.5],
+                [2.5, 2.5],
+                [3.5, 2.5],
+                [1.5, 3.5],
+                [2.5, 3.5],
+                [3.5, 3.5],
+            ],
+        ),
+        (
+            [[0, 1], [1, 0]],
+            [[-1, -1]],
+            [3, 3],
+            [
+                [-1.0, -1.0],
+                [0.0, -1.0],
+                [1.0, -1.0],
+                [-1.0, 0.0],
+                [0.0, 0.0],
+                [1.0, 0.0],
+                [-1.0, 1.0],
+                [0.0, 1.0],
+                [1.0, 1.0],
+            ],
+        ),
+        (
+            [[0, 1], [1, 0]],
+            [[10, 10]],
+            [3, 3],
+            [
+                [10.0, 10.0],
+                [11.0, 10.0],
+                [12.0, 10.0],
+                [10.0, 11.0],
+                [11.0, 11.0],
+                [12.0, 11.0],
+                [10.0, 12.0],
+                [11.0, 12.0],
+                [12.0, 12.0],
+            ],
+        ),
+        (
+            [[1, 0], [0.5, np.sqrt(3) / 2]],
+            [[0.5, 0.5 / 3**0.5], [1, 1 / 3**0.5]],
+            [2, 2],
+            [
+                [0.5, 0.28867513],
+                [1.0, 0.57735027],
+                [1.0, 1.15470054],
+                [1.5, 1.44337567],
+                [1.5, 0.28867513],
+                [2.0, 0.57735027],
+                [2.0, 1.15470054],
+                [2.5, 1.44337567],
+            ],
+        ),
+    ],
+)
+def test_positions(vectors, positions, n_cells, expected_points):
+    r"""Test that the lattice points start from the coordinates provided in the positions"""
```

Comment:
Is this checking all points or just starting points?

Reply:
This is checking all the points.

Reply:
Then the docstring is a bit misleading, or at least I could not understand it.

Reply:
How about "test that the lattice points are translated according to coordinates provided in the positions."

Reply:
Looks good.